### PR TITLE
Enable Audio sharing for screen sharing

### LINKED ISSUE #8632: Share system audio
**Is your feature request related to a problem? Please describe.**
No. This is a new feature.

**Describe the solution you'd like**
Allow presenter share your system audio.
This solution might used with only system audio(ex. Music), and/or shared screen with system audio (Ex. Play local stored video).

**Describe alternatives you've considered**
N/A

**Additional context**
N/A

### PR DESCRIPTION
**Pull request by nanovim**
### What does this PR do?
It enables a simple version of audio sharing in the html5 client.

### Closes Issue(s)

closes #8632

### Motivation
A customer wished for it and we do believe in giving code back to the community.

### More

So like I said this is HTML 5 client only and muxes the sharing stream into the already established microphone  sip connection

1. backward compatible with recording, it is still the same connection
2. for the moment muting mutes both streams
3. no changes to Freeswitch or Kurento (configuration)
4. actually a very small patch

### CODE REVIEW

**Comment by antobinary** [comment]
@nanovim Thank you for your contribution! Could you please confirm you have filled out a CLA? https://docs.bigbluebutton.org/support/faq.html#why-do-i-need-to-sign-a-contributor-license-agreement-to-contribute-source-code ?

**Comment by nanovim** [comment]
> @nanovim Thank you for your contribution! Could you please confirm you have filled out a CLA? https://docs.bigbluebutton.org/support/faq.html#why-do-i-need-to-sign-a-contributor-license-agreement-to-contribute-source-code ?

Yes I did.
I send it at this very moment.

**Comment by mrmattnewell** [comment]
@prlanzarin Hey folks, do we have any news on if this is going to get merged / added to a release soon?

**Comment by mrmattnewell** [comment]
Thanks @prlanzarin, we're super keen to see this

**Comment by ffdixon** [comment]
@nanovim Contributor agreement received - thank you.

**Comment by etherz10** [comment]
@prlanzarin Hey, do you reckon this will make it into the next 2.3 alpha?

**Comment by jmurri** [comment]
I'm also very interested in seeing this feature merged. Any updates on when this might make it into a release is much appreciated.

File: bigbluebutton-html5/imports/ui/components/screenshare/service.js [hunk 1]
```diff
@@ -58,14 +59,29 @@ const shareScreen = (onFail) => {
...
   BridgeService.getScreenStream().then((stream) => {
-    KurentoBridge.kurentoShareScreen(onFail, stream);
+    window.screenShareStream = stream;
+    //splitting up the video and audio stream into separate stream objects
+    // because Kurento does not like getting both audio and video data.
+    let videoStream = new MediaStream(stream.getVideoTracks())
+    KurentoBridge.kurentoShareScreen(onFail, videoStream)
+    if(typeof stream.getAudioTracks == "function" && stream.getAudioTracks().length > 0){
+      let audioStream = new MediaStream(stream.getAudioTracks())
+      AudioManager.joinScreenShareStream(audioStream)
```
Comment (prlanzarin) [inline bigbluebutton-html5/imports/ui/components/screenshare/service.js:69]: @nanovim have you tested this triggering a microphone reconnect? Don't think it'll pick up the audio mixer again if that happens.

Comment (nanovim) [reply bigbluebutton-html5/imports/ui/components/screenshare/service.js:69]: I actually did not think of that. 
Maybe it does, but probably not. 
I guess this needs a fix.

**Comment by mrmattnewell** [comment]
@nanovim - sorry to nudge, just so keen to see this one, any timescale on addressing @prlanzarin's review?

**Comment by mrmattnewell** [comment]
@nanovim @prlanzarin 

Hi both, I'm really keen to see this appear as soon as possible and I know that there's a great many others users that are watching this PR with interest.  I completely appreciate that you probably have other priorities both with regards to BBB and other projects.

Would it be possible to discuss us paying a commission to finish this feature? I would, of course, want the results to remain open source for the whole community.

**Comment by prlanzarin** [comment]
Yes, yes. I'll set some time aside this weekend to finish what's pending in this PR and backport it to 2.2.

**Comment by pikurasa** [comment]
I am enthused!

I teach Music Blocks (https://github.com/sugarlabs/musicblocks/) classes (which also runs in browser), and I share the audio for the kids to hear the sound, too. (I also like having the kids share their own audio as well)

I have been using Jitsi, but am looking into BBB for its scalability.

Cheers!

**Comment by mrmattnewell** [comment]
@prlanzarin - _thank you so much_.  Apologies for all the pressure, but we have big plans for BBB and this feature is crucial.  We owe you one and will hopefully be in a position to contribute significantly to the project in future.

Let me know if there's anything we can do to help.

**Comment by on8pz** [comment]
@prlanzarin You are the best!!! Our language teachers have been nagging away about this lacking feature for the past 6 months, and I was dreading having to wait for v2.3 stable to give them this feature.
Any update already ? (Sorry for being so impatient, but they will be so happy if this would work)

**Comment by prlanzarin** [comment]
I'm not sure this will go into 2.2 yet. Maybe not. I've looked at the PR this weekend, made some adjustments I still need to iron out. It'll initially go into 2.3.

The PR is going to be merged in 2.3 when it's ready. Track it for updates. Avoid pinging me for updates if possible, please.

**Comment by etherz10** [comment]
Any update on this?

**Comment by matiasilva** [comment]
@etherz10 did you read the last line of the previous comment?

**Comment by mrmattnewell** [comment]
@prlanzarin, I do appreciate that you'd asked to not be pinged, so I'm very sorry to do so. However, this feature is critical to our use of BBB (In which we have invested heavily) and we've been waiting since August for this to be reviewed and merged. As mentioned before - I'd be happy to discuss ways in which we could help with any cost associated with the time needed to get this over the line. Is there any update or an ETA of any sort?

**Comment by Jossef-767** [comment]
@prlanzarin,  I also get similar requests from many teachers who take this feature for granted and without it it is difficult for them to deliver certain types of lectures, and they are forced to use other tools

**Comment by Richardvi** [comment]
This is a request I get daily from language teachers. They use audio files during BB classes. Hope this will be fixed soon!

**Comment by mrmattnewell** [comment]
> I’ve looked at the PR this weekend, made some adjustments I still need to iron out. It’ll initially go into 2.3.

@prlanzarin / @nanovim - could you please share with me any outstanding code on this.  I'm looking to see if I can get this finished off under our own steam.

**Comment by ffdixon** [comment]
You can use the video sharing to share audio files as well.  If you upload the audio files to a dropbox account, dropbox will give you a direct URL to the file, such as

  https://dropbox.com/my/file/location.mp3

and you can use that ULR to play the audio file on everyone's computer with Share video command.

**Comment by mrmattnewell** [comment]
@ffdixon many thanks for that useful tip.  Unfortunately for our use case it's important that we can share audio from a web application while still being able to use some of the tools on the webpage - so the existing video / audio sharing capabilities don't work for us.

**Comment by Richardvi** [comment]
@ffdixon does this also work with OneDrive? Dropbox does not comply to GDPR, so we can't use it.

**Comment by ntnahn** [comment]
Hi there,
I have apply your code manually it working fine but i found a bug

Step to reproduce:
- Moderator setting the room: Mute audio when join
- Moderator: share screen with audio
- Moderator: unmute audio but got a bug => cannot heard audio + audio from share screen from other users

Please check and fix it, thank you very much.

**Comment by mrmattnewell** [comment]
@ffdixon, @nanovim, @prlanzarin 

I manage a product, so I know just how annoying these things can be, so I'm sorry to chase you - but I really need someone to engage on this one for us.  As mentioned, I'm happy to pay for someones time to assist on this, or if we can have a chat with someone, get an orientation and the latest code (@prlanzarin you mentioned above that you made some adjustments, but I can't see a commit) then I will see if I can re-task someone on our development team to have a go at making some progress.

I would very much appreciate it if someone could reply and help make a plan to get this finished off.

**Comment by prlanzarin** [comment]
I've just finished porting a different (as in a different approach) version of this to a 2.2 development branch (https://github.com/prlanzarin/bigbluebutton-1/tree/u22-screenshare18102020).

Expect a PR next week targeting 2.2.

**Comment by mrmattnewell** [comment]
@prlanzarin this is massively appreciated as I have clients shouting at me! I didn't realise that there was another separate solution in the works. Please let me know if there is anything that I can offer in assistance.

**Comment by prlanzarin** [comment]
> I didn't realise that there was another separate solution in the works.

I'll make room for this one as well so the approach is at least configurable.
Pros and cons to both approaches. Will detail it a bit better in the PR.

**Comment by prlanzarin** [comment]
See https://github.com/bigbluebutton/bigbluebutton/pull/11025. I still need to tighten the screws in a couple of rough edges there, but it's mostly done. I don't plan on lifting the draft until after 2.2.31 is released.

**Comment by tetrex** [comment]
can anyone tell me whats the status of this update? 
when will it be implemented of ETA or anything.... beta/alpha versions works too

TLDR: Whats the status of this update

**Comment by prlanzarin** [comment]
https://github.com/bigbluebutton/bigbluebutton/issues/8632

Shipped as of 2.3.0.

**Comment by ffdixon** [comment]
Closing this pull request as we've implemented sharing of audio when screen sharing on 2.3 when using Chrome.  You can share audio on a Chrome tab using mac, windows, linux, and ChromeOS.  On windows, you can also share audio when sharing the screen using Chrome.
